students/brian_warn/session04/kata_test.py

```diff
@@ -0,0 +1,94 @@
+#!/Library/Frameworks/Python.framework/Versions/3.6/bin/python
+# Kata Fourteen
+import random
+
+
+def main():
+    exceptions = ['.', '?', ' ', '!', ',', '-']
+
+    def build_trigram(infile):
+        ''' Function that builds the trigram dictionary from the source file. '''
+
+        trigrams = {}
+        with open(infile) as sherlock:
+            sherlock.readline()
+            for line in sherlock:
+                line = line.rstrip('\r\n')
+                word = line.split(" ")
+                for i in range(len(word) - 2):
+                    pair = tuple(word[i:i + 2])
+                    follower = word[i + 2]
+                    if follower not in exceptions:
+                        pass
+                    # Find dictionary method that does this in one call
+                    if pair in trigrams:
+                        trigrams[pair].append(follower)
+                    if pair not in trigrams:
+                        trigrams[tuple(pair)] = [follower]
+
+        # Check to see what's in trigrams dictionary:
+        # items = trigrams.items()
+        # for x in items:
+        #     print(x)
+        return trigrams.items()
+
+    def get_trigram_key(trigram_items, upper):
+        ''' Generates a two-word tuple. '''
+
+        # print(trigram_items)
+        len_random_choice = 1
+        # Needed since some of the random_choice results may only have one word in them.  
```

Comment:
huh?

I think I see what you are doing, but you're making it more complicated that it has to be.

the trigrams dict you created above is a gooddata structure for building new text, too.

the keys are word pairs, and the values are lists of the following words...

So to start it off, you want a random word pair, that is a random key from the dict:

`start_pair = random.choice(list(trigrams.keys()))`

you need the list() call there 'cause random.choice does not work with a dict keys object.

now you have a random pair, you can start bulding text by randomly choosinf one ofits followingwords:

`next_word = random.choice(trigrams[start_pair]))`

and away you go....

Reply:
Agreed that this is complicated.  I'll work with your recommendation and see how my code needs to change accordingly

As to where I was coming from, my approach generated the two scenarios below.  The short explanation is that I would produce results where I didn't get two words in the pair (the second run).  The first run shows that I got the two-word key that I wanted.

First Run:
```
raphael:session04 brwarn$ ./kata_test.py 
trigram_items type:  <class 'dict_items'>
random_choice in def is:  ('I', 'had')
random_choice length is:  2
trigram key is now:  ('I', 'had')
random_choice in def is:  ('Baker', 'Street.')
random_choice length is:  2
rest-of-sentence trigram is:  ('Baker', 'Street.')
```

Second Run:
```
raphael:session04 brwarn$ ./kata_test.py 
trigram_items type:  <class 'dict_items'>
random_choice in def is:  ['pacing', 'at']
random_choice length is:  2
trigram key is now:  ['pacing', 'at']
random_choice in def is:  ['now']
random_choice length is:  1
random_choice in def is:  ['associated']
random_choice length is:  1
random_choice in def is:  ('incidents', 'of')
random_choice length is:  2
rest-of-sentence trigram is:  ('incidents', 'of')
```